src/coins.h

```diff
@@ -280,11 +282,44 @@ class CCoinsViewBacked : public CCoinsView
     uint256 GetBestBlock() const override;
     std::vector<uint256> GetHeadBlocks() const override;
     void SetBackend(CCoinsView &viewIn);
-    bool BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock, bool erase = true) override;
+    bool BatchWrite(CoinsViewCacheCursor &cursor, const uint256 &hashBlock, bool erase = true) override;
     std::unique_ptr<CCoinsViewCursor> Cursor() const override;
     size_t EstimateSize() const override;
 };
 
+struct CoinsViewCacheCursor {
+public:
+    CoinsViewCacheCursor(size_t& cachedCoinsUsage, CoinsCachePair& sentinel, CCoinsMap& mapCoins) noexcept
+        : m_usage(cachedCoinsUsage), m_sentinel(sentinel), m_map(mapCoins) {}
+
+    inline CoinsCachePair* Begin() noexcept
+    {
+        m_curr = m_sentinel.second.Next();
+        return m_curr;
+    }
```

Comment:
I find it very confusing when a method mutates AND returns.
Maybe we could either do something like
```C++
    for (cursor.Begin(); cursor.GetCurrent() != cursor.End(); cursor.Next(erase)) {
        auto it = cursor.GetCurrent();
```
in which case `Begin` and `Next` modify the state but don't actually return anything.

--------

Or even better, we could get rid `CoinsCachePair* m_curr{nullptr};` (`Begin` would also become a `const`) and add it to `Next(CoinsCachePair* current, bool erase)`, and the usage would simply become `cursor.Next(it, erase)` instead.

<details>

<summary>suggested patch</summary>

```patch
diff --git a/src/coins.h b/src/coins.h
--- a/src/coins.h	(revision 04ed226236d1ef12110cb5f85d4d3f133d2f68a8)
+++ b/src/coins.h	(date 1720523816507)
@@ -292,33 +292,25 @@
     CoinsViewCacheCursor(size_t& cachedCoinsUsage, CoinsCachePair& sentinel, CCoinsMap& mapCoins) noexcept
         : m_usage(cachedCoinsUsage), m_sentinel(sentinel), m_map(mapCoins) {}
 
-    inline CoinsCachePair* Begin() noexcept
-    {
-        m_curr = m_sentinel.second.Next();
-        return m_curr;
-    }
-
+    inline CoinsCachePair* Begin() const noexcept { return m_sentinel.second.Next(); }
     inline CoinsCachePair* End() const noexcept { return &m_sentinel; }
-
-    inline CoinsCachePair* Next(bool erase) noexcept
+    inline CoinsCachePair* Next(CoinsCachePair* current, bool erase) noexcept
     {
-        const auto next_entry{m_curr->second.Next()};
+        const auto next_entry{current->second.Next()};
         if (!erase) {
-            if (m_curr->second.coin.IsSpent()) {
-                m_usage -= m_curr->second.coin.DynamicMemoryUsage();
-                m_map.erase(m_curr->first);
+            if (current->second.coin.IsSpent()) {
+                m_usage -= current->second.coin.DynamicMemoryUsage();
+                m_map.erase(current->first);
             } else {
-                m_curr->second.ClearFlags();
+                current->second.ClearFlags();
             }
         }
-        m_curr = next_entry;
-        return m_curr;
+        return next_entry;
     }
 private:
     size_t& m_usage;
     CoinsCachePair& m_sentinel;
     CCoinsMap& m_map;
-    CoinsCachePair* m_curr{nullptr};
 };
 
 /** CCoinsView that adds a memory cache for transactions to another CCoinsView */
Index: src/txdb.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/txdb.cpp b/src/txdb.cpp
--- a/src/txdb.cpp	(revision 04ed226236d1ef12110cb5f85d4d3f133d2f68a8)
+++ b/src/txdb.cpp	(date 1720523312389)
@@ -124,7 +124,7 @@
             changed++;
         }
         count++;
-        it = cursor.Next(erase);
+        it = cursor.Next(it, erase);
         if (batch.SizeEstimate() > m_options.batch_write_bytes) {
             LogPrint(BCLog::COINDB, "Writing partial batch of %.2f MiB\n", batch.SizeEstimate() * (1.0 / 1048576.0));
             m_db->WriteBatch(batch);
Index: src/coins.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/coins.cpp b/src/coins.cpp
--- a/src/coins.cpp	(revision 04ed226236d1ef12110cb5f85d4d3f133d2f68a8)
+++ b/src/coins.cpp	(date 1720523312387)
@@ -190,7 +190,7 @@
 }
 
 bool CCoinsViewCache::BatchWrite(CoinsViewCacheCursor &cursor, const uint256 &hashBlockIn, bool erase) {
-    for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.Next(erase)) {
+    for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.Next(it, erase)) {
         // Ignore non-dirty entries (optimization).
         if (!(it->second.IsDirty())) {
             continue;
Index: src/test/fuzz/coinscache_sim.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/fuzz/coinscache_sim.cpp b/src/test/fuzz/coinscache_sim.cpp
--- a/src/test/fuzz/coinscache_sim.cpp	(revision 04ed226236d1ef12110cb5f85d4d3f133d2f68a8)
+++ b/src/test/fuzz/coinscache_sim.cpp	(date 1720523312388)
@@ -174,7 +174,7 @@
 
     bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256&, bool erase) final
     {
-        for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.Next(erase)) {
+        for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.Next(it, erase)) {
             if (it->second.IsDirty()) {
                 if (it->second.coin.IsSpent() && (it->first.n % 5) != 4) {
                     m_data.erase(it->first);
Index: src/test/coins_tests.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/coins_tests.cpp b/src/test/coins_tests.cpp
--- a/src/test/coins_tests.cpp	(revision 04ed226236d1ef12110cb5f85d4d3f133d2f68a8)
+++ b/src/test/coins_tests.cpp	(date 1720523312388)
@@ -57,7 +57,7 @@
 
     bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hashBlock, bool erase = true) override
     {
-        for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.Next(erase)){
+        for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.Next(it, erase)){
             if (it->second.IsDirty()) {
                 // Same optimization used in CCoinsViewDB is to only write dirty entries.
                 map_[it->first] = it->second.coin;

```

</details>

Reply:
Done, thanks!